### PR TITLE
Add filter on execution list

### PR DESCRIPTION
**Pull request by antho1404**
Add filter for execution list on the keeper.

Filter available:
- parentHash: base58 hash
- eventHash: base58 hash
- instanceHash: base58 hash
- processHash: base58 hash
- status string (Proposed, InProgress, Completed, Failed)
- stepKey: string

These filters are available as query parameter in the `/execution/list` API 
eg:
```
curl http://localhost:1317/execution/list\?instanceHash\=7dTayprXsBTzc7Z6iqCTjfJzh3Sbtasdgh9wdPzT4gUL\&processHash\=BACectqn5sHeDfK7N6dQdCKKY9ZfrCt5Rh428RYERYXU
```

### CODE REVIEW

File: x/execution/internal/types/querier.go [hunk 1]
```diff
@@ -1,6 +1,30 @@
...
+	Status       execution.Status `json:"status"`
+}
+
+// Match returns true if an execution matches a specific filter
+func (f ListFilter) Match(exec *execution.Execution) bool {
+	return (f.Status == execution.Status_Unknown || f.Status == exec.Status) &&
+		(f.ProcessHash.IsZero() || f.ProcessHash.Equal(exec.ProcessHash)) &&
+		(f.InstanceHash.IsZero() || f.InstanceHash.Equal(exec.InstanceHash)) &&
+		(f.ParentHash.IsZero() || f.ParentHash.Equal(exec.ParentHash)) &&
+		(f.EventHash.IsZero() || f.EventHash.Equal(exec.EventHash))
```
Comment (NicolasMahe) [inline x/execution/internal/types/querier.go:29]: this is really hard to read
maybe a longer but easier to read if statements will be nicer?

File: x/execution/client/rest/query.go [hunk 2]
```diff
@@ -62,7 +63,57 @@ func queryListHandlerFn(cliCtx context.CLIContext) http.HandlerFunc {
...
+		if param := r.FormValue("processHash"); param != "" {
+			h, err := hash.Decode(param)
+			if err != nil {
+				rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
+				return
+			}
+			filter.ProcessHash = h
+		}
+
+		if param := r.FormValue("status"); param != "" {
```
Comment (NicolasMahe) [inline x/execution/client/rest/query.go:103]: is `status` a number or a string in the request?

Comment (antho1404) [reply x/execution/client/rest/query.go:103]: This is a string, I updated the description

File: x/execution/client/rest/query.go [hunk 3]
```diff
@@ -62,7 +63,61 @@ func queryListHandlerFn(cliCtx context.CLIContext) http.HandlerFunc {
...
+			status, ok := execution.Status_value[param]
+			if !ok {
+				rest.WriteErrorResponse(w, http.StatusBadRequest, fmt.Sprintf("%q is invalid", param))
```
Comment (NicolasMahe) [inline x/execution/client/rest/query.go:106]:
```suggestion
				rest.WriteErrorResponse(w, http.StatusBadRequest, "error on parameter status: value is invalid")
```

File: x/execution/client/rest/query.go [hunk 4]
```diff
@@ -62,7 +63,61 @@ func queryListHandlerFn(cliCtx context.CLIContext) http.HandlerFunc {
...
+			filter.Status = execution.Status(status)
+		}
+
+		if param := r.FormValue("nodeKey"); param != "" {
+			filter.NodeKey = param
+		}
+
+		data, err := cliCtx.Codec.MarshalJSON(filter)
+		if err != nil {
+			return
```
Comment (NicolasMahe) [inline x/execution/client/rest/query.go:118]:
```suggestion
			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
			return
```

**Pull request merged**